New hitreg system

notdap opened the pull request:
# Description

Added a new option: "Enable Improved Hit Register" to use my hit register patches. This new hit register system is lag compensated and improves on some details such as player-player vision reach reducing. The changes are mostly based on Hypixel's changes which makes hit register fair for absolutely everyone regardless of the ping. The lag compensator basically stores the locations of the players every x milliseconds so we can retrieve the position of a victim based on the ping of the attacker and "check" if the hit was actually supposed to go through in the attacker's perspective.

# How has this been tested?

The changes have been tested on a local server and on a server I own. They work absolutely fine without any issues.

# Additional Comments

The lag compensator is a port of the following plugin: https://github.com/Islandscout/ServerSideHitDetection
My changes are based on what Hypixel has let its players know through this blog post: https://hypixel.net/threads/dev-blog-2-changes-to-hit-registration-your-feedback.1497403/

# Checklist:

- [x] I have reviewed my code thoroughly.
- [x] I have tested my code.
- [x] My changes generate no new warnings

sadcenter commented on NachoSpigot-Server/src/main/java/dev/cobblesword/nachospigot/hitdetection/LagCompensator.java:
```diff
@@ -0,0 +1,96 @@
+        Adjusts hit detection to be more fair for players with worse ping.
+     */
+
+    private static final Map<UUID, List<Pair<Location, Long>>> locationTimes = new HashMap<>();
```
you can use MultiMap here

sadcenter replied:
why don't you use MultiMap here?

sadcenter commented on NachoSpigot-Server/src/main/java/dev/cobblesword/nachospigot/hitdetection/LagCompensator.java:
```diff
@@ -0,0 +1,96 @@
+    private static final Map<UUID, List<Pair<Location, Long>>> locationTimes = new HashMap<>();
+    private static final int historySize = 35;
+    private static final int pingOffset = 175;
+    private static final int TIME_RESOLUTION = 30;
```
constants names should be like that:
HISTORY_SIZE
PING_OFFSET
TIME_RESOLUTION

sadcenter commented on NachoSpigot-Server/src/main/java/net/minecraft/server/EntityLiving.java:
```diff
@@ -1789,7 +1790,34 @@ public void receive(Entity entity, int i) {
+                    new Vec3D(entity.locX, entity.locY + (parts * 1), entity.locZ)
+            ) == null;
+
+            return firstCondition | secondCondition | thirdCondition;
```
it will execute all 3 checks even if one is already true so it will return true anyway (also, didn't you want to use `||`?) 
```suggestion
        return this.world.rayTrace(
                new Vec3D(this.locX, this.locY + (double) this.getHeadHeight(), this.locZ),
                new Vec3D(entity.locX, entity.locY + (parts * 3), entity.locZ)
        ) == null || this.world.rayTrace(
                new Vec3D(this.locX, this.locY + (double) this.getHeadHeight(), this.locZ),
                new Vec3D(entity.locX, entity.locY + (parts * 2), entity.locZ)
        ) == null || this.world.rayTrace(
                new Vec3D(this.locX, this.locY + (double) this.getHeadHeight(), this.locZ),
                new Vec3D(entity.locX, entity.locY + (parts * 1), entity.locZ)
        ) == null;
```

sadcenter commented on NachoSpigot-Server/src/main/java/dev/cobblesword/nachospigot/hitdetection/Pair.java:
```diff
@@ -0,0 +1,25 @@
+package dev.cobblesword.nachospigot.hitdetection;
+
+public class Pair<K,V> {
```
Pair is already in apache, so you can use it

sadcenter reviewed (changes requested):
It would be better to make `LagCompensator` an object and put it in `MinecraftServer` with getters etc.

notdap commented:
Just updated it all, thanks for the review! I must say many of those "issues" are because I didn't notice them while porting over the Lag Compensator. I've fixed it all based on your review but there might now be some more issues, since this is my first time with this kind of stuff.

Lucaskyy commented on NachoSpigot-Server/src/main/java/net/minecraft/server/EntityLiving.java:
```diff
@@ -1789,7 +1790,28 @@ public void receive(Entity entity, int i) {
+
+            double parts = entity.getHeadHeight() / 3;
+
+            return this.world.rayTrace(
```
This looks fine to me, but was this "benchmarked" in some way? Is it rare for the 3 ray traces to get called, or do they all get called every time? Since then this might get expensive. I'm not sure, you can ignore this comment for now.

notdap replied:
No, I haven't benchmarked it. From a really general viewpoint, the new changes do not affect the server performance at all. I've been running an older patch for this hit detection system with some optimization issues and it runs absolutely fine. Didn't cause any lag but I personally haven't tested it with higher player counts.

Lucaskyy commented on NachoSpigot-Server/src/main/java/net/minecraft/server/PlayerConnection.java:
```diff
@@ -1493,10 +1495,11 @@ public void a(PacketPlayInUseEntity packetplayinuseentity) {
             if (!flag) {
-                d0 = 9.0D;
+                // Increase the no player-player vision maximum reach
+                d0 = (NachoConfig.enableImprovedHitReg) ? 12.75D : 9.0D;
```
I can't see what code is above this line, but does this mess with anticheats or not? Might need to be tested.

notdap replied:
The server I run utilizes the Vulcan anti-cheat and I haven't run into any issues yet, although it might actually break some things depending on the anti-cheat. I personally haven't checked to see if the checks have been altered in any way, so I can't say for sure. But I can definitely say that my changes have not caused any false positives yet.

Lucaskyy reviewed (commented):
Everything looks good to me, I've got a couple of questions and notes but besides that LGTM! Well done, appreciate it :)

notdap commented on NachoSpigot-Server/src/main/java/dev/cobblesword/nachospigot/hitdetection/LagCompensator.java:
```diff
@@ -0,0 +1,95 @@
+            int elapsedTime = (int) (currentTime - times.get(i).getValue());
+
+            if (elapsedTime >= rewindTime) {
+                if (i == times.size() - 1)
```
Just added that, thanks for letting me know!

Lucaskyy commented on NachoSpigot-Server/src/main/java/dev/cobblesword/nachospigot/hitdetection/LagCompensator.java:
```diff
@@ -37,12 +37,13 @@ public Location getHistoryLocation(int rewindMillisecs, Player player) {
         long currentTime = System.currentTimeMillis();
 
         int rewindTime = rewindMillisecs + pingOffset;
+        int timesSize = times.size();
```
One small thing, you never use `timesSize`, you always use `timesSize - 1`. I'd recommend changing this line to `int timesSize = times.size() - 1` and remove the calculations in the for loop.

notdap replied:
Not sure how I missed that but I just fixed it. Thanks for letting me know about it

Lucaskyy commented on NachoSpigot-Server/src/main/java/dev/cobblesword/nachospigot/hitdetection/LagCompensator.java:
```diff
@@ -37,12 +37,13 @@ public Location getHistoryLocation(int rewindMillisecs, Player player) {
+        int timesSize = times.size();
 
-        for (int i = times.size() - 1; i >= 0; i--) {
+        for (int i = timesSize - 1; i >= 0; i--) {
```
Here

Lucaskyy commented on NachoSpigot-Server/src/main/java/dev/cobblesword/nachospigot/hitdetection/LagCompensator.java:
```diff
@@ -37,12 +37,13 @@ public Location getHistoryLocation(int rewindMillisecs, Player player) {
 
             if (elapsedTime >= rewindTime) {
-                if (i == times.size() - 1)
+                if (i == timesSize - 1)
```
And here

Elierrr commented on NachoSpigot-Server/src/main/java/net/minecraft/server/MinecraftServer.java:
```diff
@@ -118,6 +119,8 @@
     public int autosavePeriod;
     // CraftBukkit end
 
+    private final LagCompensator lagCompensator;
```
This should be moved to Nacho.java

Elierrr commented on NachoSpigot-Server/src/main/java/net/minecraft/server/MinecraftServer.java:
```diff
@@ -138,6 +141,8 @@ public MinecraftServer(OptionSet options, Proxy proxy, File file1) {
             Main.useJline = false;
         }
 
+        this.lagCompensator = new LagCompensator(35, 175, 30);
```
Move to Nacho init method

Elierrr commented on NachoSpigot-Server/src/main/java/net/minecraft/server/MinecraftServer.java:
```diff
@@ -1120,6 +1125,10 @@ public String getServerModName() {
         return NachoConfig.serverBrandName; // [Nacho-0035] // NachoSpigot - NachoSpigot >  // TacoSpigot - TacoSpigot // PaperSpigot - PaperSpigot > // Spigot - Spigot > // CraftBukkit - cb > vanilla!
     }
 
+    public LagCompensator getLagCompensator() {
```
Move to Nacho.java

Lucaskyy commented on NachoSpigot-Server/src/main/java/dev/cobblesword/nachospigot/hitdetection/LagCompensator.java:
```diff
@@ -0,0 +1,96 @@
+        long currentTime = System.currentTimeMillis();
+
+        int rewindTime = rewindMillisecs + pingOffset;
+        int timesSize = times.size() - 1;
```
Question, is `times.size()` accessed from multiple threads? If so, watch out for data races. They are a pain in the ass to bug-find. If so, look for something thread-safe. If not, keep it like this :)

notdap replied:
I'm not precisely sure, since I'm not fully aware of how threading is done in the server. This method is only called when the server receives a Entity Use packet through a method from the Entity class.

Lucaskyy replied:
Then this is fine for now. If async packets may be added in the future we will have to keep this in mind.

Elierrr commented on NachoSpigot-Server/src/main/java/dev/cobblesword/nachospigot/hitdetection/LagCompensator.java:
```diff
@@ -0,0 +1,96 @@
+        this.timeResolution = timeResolution;
+    }
+
+    public Location getHistoryLocation(int rewindMillisecs, Player player) {
```
This method is always called with the first arg being the players ping so this should be changed to remove the rewindMillisecs arg and make a variable for rewindMillisecs with the players ping. (`player.spigot().getPing()`) This also resolves Sculas’s comment about the cast optimization.

Lucaskyy commented:
I'm happy with the current state of this PR. I'll allow some of the others to have their say too, so unless there are no responses till tomorrow, I'll merge it :)

Elierrr reviewed (commented):
This review isn’t very in-depth and are just minor changes, I’ll do an in-depth review when I have time.

notdap commented:
Just changed the few things requested, sorry for being a bit late

demkom58 commented on NachoSpigot-Server/src/main/java/net/minecraft/server/Entity.java:
```diff
@@ -1181,11 +1183,24 @@ public double f(double d0, double d1, double d2) {
-        double d2 = this.locZ - entity.locZ;
+        if (entity instanceof EntityPlayer && NachoConfig.enableImprovedHitReg) {
+            Location loc = MinecraftServer.getServer().getLagCompensator().getHistoryLocation(
+                    ((EntityPlayer) entity).ping,
```
It will be better to do config option check first, because instanceof is slower than just boolean check.

Lucaskyy commented on NachoSpigot-Server/src/main/java/net/minecraft/server/Entity.java:
```diff
@@ -1181,11 +1184,23 @@ public double f(double d0, double d1, double d2) {
-        double d0 = this.locX - entity.locX;
-        double d1 = this.locY - entity.locY;
-        double d2 = this.locZ - entity.locZ;
+        if (entity instanceof EntityPlayer && NachoConfig.enableImprovedHitReg) {
```
Boolean compare is less expensive than the `instanceof` operator. (thank you for noticing @demkom58!)
```suggestion
        if (NachoConfig.enableImprovedHitReg && entity instanceof EntityPlayer) {
```

Lucaskyy commented:
Thanks a lot for the PR @notdap and thanks to everyone who helped with this PR! :)

HeathLoganCampbell commented on NachoSpigot-Server/src/main/java/dev/cobblesword/nachospigot/hitdetection/LagCompensator.java:
```diff
@@ -0,0 +1,79 @@
+
+        int rewindMillisecs = player.spigot().getPing();
+
+        List<Pair<Location, Long>> times = locationTimes.get(player.getUniqueId());
```
Personally, I would probably make this a class instead of a pair, I would probably call it LocationSnapshot, and state the double x, y, z; long registeredTimestamp; so we don't hold that heavy world reference by mistake somehow.

Also for the variable name, I would probably call it "previousLocations"

HeathLoganCampbell commented on NachoSpigot-Server/src/main/java/dev/cobblesword/nachospigot/hitdetection/LagCompensator.java:
```diff
@@ -0,0 +1,79 @@
+    private final int pingOffset = 92;
+    private final int timeResolution = 30;
+
+    public Location getHistoryLocation(Player player) {
```
Might be worth adding a comment here explaining how it works, like "Gets the closest location to the attackers ping" or something.

notdap replied:
This made me realize after one of the changes it broke the "compensator". This method used to have another parameter used to specify the ping of the attacker, and it's now instead using the ping of the victim, so that's quite useless. I'm working on fixing this atm

HeathLoganCampbell commented on NachoSpigot-Server/src/main/java/dev/cobblesword/nachospigot/hitdetection/LagCompensator.java:
```diff
@@ -0,0 +1,79 @@
+        int timesSize = times.size() - 1;
+
+        for (int i = timesSize; i >= 0; i--) {
+            int elapsedTime = (int) (currentTime - times.get(i).getValue());
```
`times.get(i)` is used a few times here, maybe try use a variable

HeathLoganCampbell commented on NachoSpigot-Server/src/main/java/dev/cobblesword/nachospigot/hitdetection/LagCompensator.java:
```diff
@@ -0,0 +1,79 @@
+                if (i == timesSize)
+                    return times.get(i).getKey();
+
+                double nextMoveWeight = (elapsedTime - rewindTime) / (double) (elapsedTime - (currentTime - times.get(i + 1).getValue()));
```
This should probably be broken down to add context

```
int maxRewindMilli = rewindMillisecs + pingOffset; // Instead of rewindTime 

// ....
int millisSinceLoc = (int) (currentTime - times.get(i).getValue()); // instead of elapsedTime 

double movementRelAge = millisSinceLoc - maxRewindMilli;
double millisSinceLastLoc = currentTime - times.get(i + 1).getValue()
double nextMoveWeight = movementRelAge / (millisSinceLoc - millisSinceLastLoc)
```

HeathLoganCampbell commented on NachoSpigot-Server/src/main/java/me/elier/nachospigot/config/NachoConfig.java:
```diff
@@ -390,4 +391,11 @@ private static void enableProtocolShim() {
+    public static boolean enableImprovedHitReg;
+
+    private static void enableImprovedHitReg() {
+        enableImprovedHitReg = getBoolean("settings.use-improved-hitreg", false);
```
I like how you have it disabled by default, It is a pretty high risk change, nice job

notdap replied:
Yeah this hitreg can be very different from what people are used to so I thought not everyone would be interested in having it on

HeathLoganCampbell commented on NachoSpigot-Server/src/main/java/net/minecraft/server/EntityLiving.java:
```diff
@@ -1789,7 +1790,28 @@ public void receive(Entity entity, int i) {
-        return this.world.rayTrace(new Vec3D(this.locX, this.locY + (double) this.getHeadHeight(), this.locZ), new Vec3D(entity.locX, entity.locY + (double) entity.getHeadHeight(), entity.locZ)) == null;
+        if (entity instanceof EntityPlayer && NachoConfig.enableImprovedHitReg) {
+            // Head height is 1,5725
+            // Split it into three to get a more accurate line of sight -> 0.52416667
```
Nice comment

HeathLoganCampbell commented on NachoSpigot-Server/src/main/java/net/minecraft/server/EntityLiving.java:
```diff
@@ -1789,7 +1790,28 @@ public void receive(Entity entity, int i) {
+                    new Vec3D(this.locX, this.locY + (double) this.getHeadHeight(), this.locZ),
+                    new Vec3D(entity.locX, entity.locY + (parts * 3), entity.locZ)
+            ) == null || this.world.rayTrace(
+                    new Vec3D(this.locX, this.locY + (double) this.getHeadHeight(), this.locZ),
```
avoid creating unneed objects here  new Vec3D(this.locX, this.locY + (double) this.getHeadHeight(), this.locZ) can be reused

HeathLoganCampbell commented on NachoSpigot-Server/src/main/java/net/minecraft/server/EntityLiving.java:
```diff
@@ -1789,7 +1790,28 @@ public void receive(Entity entity, int i) {
+            // Head height is 1,5725
+            // Split it into three to get a more accurate line of sight -> 0.52416667
+
+            double parts = entity.getHeadHeight() / 3;
```
It's kinda iffy dividing by 3 and calling ray trace, as that's one of the most expensive calls you can make, is there any other solutions here?

notdap replied:
I personally don't know of any other ways to do it

HeathLoganCampbell commented on NachoSpigot-Server/src/main/java/net/minecraft/server/PlayerList.java:
```diff
@@ -574,6 +575,9 @@ public EntityPlayer moveToWorld(EntityPlayer entityplayer, int i, boolean flag,
 
             Player respawnPlayer = cserver.getPlayer(entityplayer1);
+
+            Nacho.get().getLagCompensator().registerMovement(respawnPlayer, location);
```
On respawn, the players history should be cleared, as that's dead data at this point

HeathLoganCampbell reviewed (commented):
Very interesting change dude, would love to see more from you in the future :o
